components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/graphQL/GraphQLQueryAnalysisHandler.java

```diff
@@ -0,0 +1,353 @@
+            }
+        } catch (ParseException e) {
+            log.error("Policy definition parsing failed. " + e.getMessage(), e);
+        }
+        return false;
+    }
+
+    /**
+     * This method analyses the query depth
+     *
+     * @param messageContext   message context of the request
+     * @param payload          payload of the request
+     * @param policyDefinition json object which contains the policy
+     * @return true, if the query depth does not exceed the maximum value or false, if query depth exceeds the maximum
+     */
+    private boolean analyseQueryDepth(MessageContext messageContext, String payload, JSONObject policyDefinition) {
+        String username = APISecurityUtils.getAuthenticationContext(messageContext).getUsername();
+
+        try {
+            String[] userRoles = getUserRoles(username);
+            int maxQueryDepth = getMaxQueryDepth(userRoles, policyDefinition);
+            if (maxQueryDepth > 0) {
+                MaxQueryDepthInstrumentation maxQueryDepthInstrumentation =
+                        new MaxQueryDepthInstrumentation(maxQueryDepth);
+                GraphQL runtime = GraphQL.newGraphQL(schema).instrumentation(maxQueryDepthInstrumentation).build();
+
+                try {
+                    ExecutionResult executionResult = runtime.execute(payload);
+                    List<GraphQLError> errors = executionResult.getErrors();
+                    if (errors.size() > 0) {
+                        List<String> errorList = new ArrayList<>();
+                        for (GraphQLError error : errors) {
+                            errorList.add(error.getMessage());
+                        }
+
+                        // TODO: https://github.com/wso2/carbon-apimgt/issues/8147
+                        ListIterator<String> iterator = errorList.listIterator();
+                        while (iterator.hasNext()) {
+                            if (iterator.next().contains("non-nullable")) {
+                                iterator.remove();
+                            }
+                        }
+
+                        if (errorList.size() == 0) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Maximum query depth of " + maxQueryDepth + " was not exceeded");
+                            }
+                            return true;
+                        }
+
+                        handleFailure(APISecurityConstants.GRAPHQL_QUERY_TOO_DEEP, messageContext,
+                                APISecurityConstants.GRAPHQL_QUERY_TOO_DEEP_MESSAGE, errorList.toString());
+                        log.error(errorList.toString());
+                        return false;
+                    }
+                    if (log.isDebugEnabled()) {
+                        log.debug("Maximum query depth of " + maxQueryDepth + " was not exceeded");
+                    }
+                    return true;
+                } catch (Throwable e) {
```

Comment:
can't we catch specific exception here ?

Reply:
Fixed with https://github.com/wso2/carbon-apimgt/pull/10929